lib/src/rules/parameter_assignments.dart

```diff
@@ -124,75 +106,95 @@ class ParameterAssignments extends LintRule {
   }
 }
 
-class _Visitor extends SimpleAstVisitor<void> {
+class _DeclarationVisitor extends RecursiveAstVisitor {
+  final FormalParameter parameter;
   final LintRule rule;
+  final bool paramIsNotNullByDefault;
+  final bool paramDefaultsToNull;
+  bool hasBeenAssigned = false;
 
-  _Visitor(this.rule);
+  _DeclarationVisitor(this.parameter, this.rule)
+      : paramIsNotNullByDefault = parameter is SimpleFormalParameter ||
+            _isDefaultFormalParameterWithDefaultValue(parameter),
+        paramDefaultsToNull = parameter is DefaultFormalParameter &&
+            parameter.defaultValue == null;
 
   @override
-  void visitFunctionDeclaration(FunctionDeclaration node) {
-    var parameters = node.functionExpression.parameters;
-    if (parameters != null) {
-      // Getter do not have formal parameters.
-      for (var e in parameters.parameters) {
-        var declaredElement = e.declaredElement;
-        if (declaredElement != null &&
-            node.functionExpression.body
-                .isPotentiallyMutatedInScope(declaredElement)) {
-          _reportIfSimpleParameterOrWithDefaultValue(e, node);
+  visitAssignmentExpression(AssignmentExpression node) {
+    if (paramIsNotNullByDefault) {
+      if (_isFormalParameterReassigned(parameter, node)) {
+        rule.reportLint(node);
+      }
+    } else {
+      if (paramDefaultsToNull) {
```

Comment:
nit: I'd make this an `else if` in order to show that the condition is on the same level as (or parallel to) `paramIsNotNullByDefault`; that is, the outer conditions are testing the nature of the parameter whereas inner conditions are testing the assignment.